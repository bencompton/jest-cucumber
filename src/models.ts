import { Options } from './configuration';

export type StepFromStepDefinitions = {
    stepMatcher: string | RegExp;
    stepFunction(stepArguments?: any): void | PromiseLike<any>;
};

export type ScenarioFromStepDefinitions = {
    title: string;
    steps: StepFromStepDefinitions[];
};

export type FeatureFromStepDefinitions = {
    title: string;
    scenarios: ScenarioFromStepDefinitions[];
};

export type ParsedStep = {
    keyword: string;
    stepText: string;
    stepArgument: string | {};
    lineNumber: number;
};

export type ParsedScenario = {
    title: string;
    steps: ParsedStep[];
    tags: string[];
    lineNumber: number;
    skippedViaTagFilter: boolean;
};

export type ParsedScenarioOutline = {
    title: string;
    tags: string[];
    scenarios: ParsedScenario[];
    steps: ParsedStep[];
    lineNumber: number;
    skippedViaTagFilter: boolean;
};

export type ParsedFeature = {
    title: string;
    scenarios: ParsedScenario[];
    scenarioOutlines: ParsedScenarioOutline[];
    options: Options;
    tags: string[];
};
