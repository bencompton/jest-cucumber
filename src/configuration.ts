export type ErrorOptions = {
    scenariosMustMatchFeatureFile: boolean;
    stepsMustMatchFeatureFile: boolean;
    allowScenariosNotInFeatureFile: boolean;
};

export type Options = {
    loadRelativePath?: boolean;
    tagFilter?: string;
    errors?: ErrorOptions | boolean;
    scenarioNameTemplate?: (vars: ScenarioNameTemplateVars) => string;
};

export type ScenarioNameTemplateVars = {
    featureTitle: string;
    scenarioTitle: string;
    scenarioTags: string[];
    featureTags: string[];
};

export const defaultErrorSettings = {
    scenariosMustMatchFeatureFile: true,
    stepsMustMatchFeatureFile: true,
    allowScenariosNotInFeatureFile: false,
};

export const defaultConfiguration: Options = {
    tagFilter: undefined,
    scenarioNameTemplate: undefined,
    errors: defaultErrorSettings,
};

let globalConfiguration: Options = {} as Options;

export const getJestCucumberConfiguration = (options?: Options) => {
    const mergedOptions = { ...defaultConfiguration, ...globalConfiguration, ...options || {} };

    if (mergedOptions.errors === true) {
        mergedOptions.errors = defaultErrorSettings;
    }

    return mergedOptions;
};

export const setJestCucumberConfiguration = (options: Options) => {
    globalConfiguration = options;
};
