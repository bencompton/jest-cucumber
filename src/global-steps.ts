import { StepFromStepDefinitions } from './models';

class Steps {
    list: StepFromStepDefinitions[] = [];

    push(step: StepFromStepDefinitions) {
        const { stepMatcher } = step;

        const isAlreadyAdded = this.list.some(
            step => String(step.stepMatcher) === String(stepMatcher),
        );

        if (isAlreadyAdded) {
            throw new Error(`Existing global step with the name ${stepMatcher}`);
        }

        this.list.push(step);
    }

    get(title: StepFromStepDefinitions["stepMatcher"]) {
        let params;

        const found = this.list.find((step) => {
            if (typeof title === 'string') {
                const matches = title.match(step.stepMatcher);

                if (matches) {
                    params = matches.slice(1);
                    return true;
                }
            }

            if (title instanceof RegExp) {
                return String(title) === String(step.stepMatcher);
            }

            return false;
        });

        if (!found) {
            throw Error(`${title} : was not defined in Steps file`);
        }

        return found.stepFunction.bind(this, ...params);
    }
}

export const globalSteps = new Steps();

export function defineGlobalStep(
    stepMatcher: StepFromStepDefinitions["stepMatcher"],
    stepFunction: StepFromStepDefinitions["stepFunction"]
) {
    globalSteps.push({ stepMatcher, stepFunction });
}
