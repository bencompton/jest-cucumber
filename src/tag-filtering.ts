import { ParsedFeature, ParsedScenario, ParsedScenarioOutline } from './models';

type TagFilterFunction = (tags: string[]) => boolean;

const cachedTagFilterFunctions: { [tag: string]: TagFilterFunction } = {};

const convertTagFilterExpressionToFunction = (tagFilterExpression: string) => {
  const tagRegex = /(@[A-Za-z-_.0-9]+)/g;
  const tags: string[] = [];
  let match: RegExpMatchArray | null = null;
  let newTagFilterExpression = `${tagFilterExpression}`;

  do {
    match = tagRegex.exec(tagFilterExpression);

    if (match) {
      newTagFilterExpression = newTagFilterExpression.replace(
        match[1],
        `(tags.indexOf("${match[1].toLowerCase()}")!==-1)`,
      );

      if (tags.indexOf(match[1]) !== -1) {
        tags.push(match[1]);
      }
    }
  } while (match);

  newTagFilterExpression = newTagFilterExpression.replace(/(\s+not|not\s+|\s+not\s+)/g, ' ! ');
  newTagFilterExpression = newTagFilterExpression.replace(/(\s+or|or\s+|\s+or\s+)/g, ' || ');
  newTagFilterExpression = newTagFilterExpression.replace(/(\s+and|and\s+|\s+and\s+)/g, ' && ');
  newTagFilterExpression = newTagFilterExpression.replace(/[ \t\n\r]+/g, '');

  let tagFilterFunction: TagFilterFunction;

  try {
    // eslint-disable-next-line no-new-func
    tagFilterFunction = new Function('tags', `return ${newTagFilterExpression};`) as TagFilterFunction;
    tagFilterFunction([]);
  } catch (error) {
    throw new Error(`Could not parse tag filter "${tagFilterExpression}"`);
  }

  return tagFilterFunction;
};

const checkIfScenarioMatchesTagFilter = (
  tagFilterExpression: string,
  feature: ParsedFeature,
  scenario: ParsedScenario | ParsedScenarioOutline,
) => {
  const featureAndScenarioTags = [
    ...scenario.tags.map(tag => tag.toLowerCase()),
    ...feature.tags.map(tag => tag.toLowerCase()),
  ];

  let tagFilterFunction = cachedTagFilterFunctions[tagFilterExpression];

  if (!tagFilterFunction) {
    tagFilterFunction = convertTagFilterExpressionToFunction(tagFilterExpression);
    cachedTagFilterFunctions[tagFilterExpression] = tagFilterFunction;
  }

  return tagFilterFunction(featureAndScenarioTags);
};

const setScenarioSkipped = (parsedFeature: ParsedFeature, scenario: ParsedScenario) => {
  const skippedViaTagFilter = !checkIfScenarioMatchesTagFilter(
    parsedFeature.options.tagFilter as string,
    parsedFeature,
    scenario,
  );

  return {
    ...scenario,
    skippedViaTagFilter,
  };
};

export const applyTagFilters = (parsedFeature: ParsedFeature) => {
  if (parsedFeature.options.tagFilter === undefined) {
    return parsedFeature;
  }

  const scenarios = parsedFeature.scenarios.map(scenario => setScenarioSkipped(parsedFeature, scenario));
  const scenarioOutlines = parsedFeature.scenarioOutlines.map(scenarioOutline => {
    return {
      ...setScenarioSkipped(parsedFeature, scenarioOutline),
      scenarios: scenarioOutline.scenarios.map(scenario => setScenarioSkipped(parsedFeature, scenario)),
    };
  });

  return {
    ...parsedFeature,
    scenarios,
    scenarioOutlines,
  } as ParsedFeature;
};
