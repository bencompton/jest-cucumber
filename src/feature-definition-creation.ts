import { checkThatFeatureFileAndStepDefinitionsHaveSameScenarios } from './validation/scenario-validation';
import {
    ScenarioFromStepDefinitions,
    FeatureFromStepDefinitions,
    StepFromStepDefinitions,
    ParsedFeature,
    ParsedScenario,
    ParsedScenarioOutline,
} from './models';
import {
    ensureFeatureFileAndStepDefinitionScenarioHaveSameSteps,
    matchSteps,
} from './validation/step-definition-validation';
import { applyTagFilters } from './tag-filtering';
import { Options } from './configuration';

export type StepsDefinitionCallbackOptions = {
    defineStep: DefineStepFunction;
    given: DefineStepFunction;
    when: DefineStepFunction;
    then: DefineStepFunction;
    and: DefineStepFunction;
    but: DefineStepFunction;
    pending: () => void;
};

export interface IJestLike {
    describe: jest.Describe;
    test: jest.It;
}

export type ScenariosDefinitionCallbackFunction = (defineScenario: DefineScenarioFunctionWithAliases) => void;

export type DefineScenarioFunction = (
    scenarioTitle: string,
    stepsDefinitionCallback: StepsDefinitionCallbackFunction,
    timeout?: number,
) => void;

export type DefineScenarioFunctionWithAliases = DefineScenarioFunction & {
    skip: DefineScenarioFunction;
    only: DefineScenarioFunction;
    concurrent: DefineScenarioFunction;
};

export type StepsDefinitionCallbackFunction = (options: StepsDefinitionCallbackOptions) => void;
export type DefineStepFunction = (stepMatcher: string | RegExp, stepDefinitionCallback: (...args: any[]) => any) => any;
export type DefineFeatureFunction = (
    featureFromFile: ParsedFeature,
    scenariosDefinitionCallback: ScenariosDefinitionCallbackFunction,
) => void;

export const createDefineFeature = (jestLike: IJestLike): DefineFeatureFunction => {
    const processScenarioTitleTemplate = (
        scenarioTitle: string,
        parsedFeature: ParsedFeature,
        options: Options,
        parsedScenario: ParsedScenario,
        parsedScenarioOutline: ParsedScenarioOutline,
    ) => {
        if (options && options.scenarioNameTemplate) {
            try {
                return options && options.scenarioNameTemplate({
                    featureTitle: parsedFeature.title,
                    scenarioTitle: scenarioTitle.toString(),
                    featureTags: parsedFeature.tags,
                    scenarioTags: (parsedScenario || parsedScenarioOutline).tags,
                });
            } catch (err) {
                throw new Error(
                    // tslint:disable-next-line:max-line-length
                    `An error occurred while executing a scenario name template. \nTemplate:\n${options.scenarioNameTemplate}\nError:${err.message}`,
                );
            }
        }

        return scenarioTitle;
    };

    const checkForPendingSteps = (scenarioFromStepDefinitions: ScenarioFromStepDefinitions) => {
        let scenarioPending = false;

        scenarioFromStepDefinitions.steps.forEach((step) => {
            try {
                if (step.stepFunction.toString().indexOf('pending()') !== -1) {
                    const pendingTest = new Function(`
                        let isPending = false;

                        const pending = function () {
                            isPending = true;
                        };

                        (${step.stepFunction})();

                        return isPending;
                    `);

                    scenarioPending = pendingTest();
                }
            } catch (err) {
                // Ignore
            }
        });

        return scenarioPending;
    };

    const getTestFunction = (skippedViaTagFilter: boolean, only: boolean, skip: boolean, concurrent: boolean) => {
        if (skip || skippedViaTagFilter) {
            return jestLike.test.skip;
        } else if (only) {
            return jestLike.test.only;
        } else if (concurrent) {
            return jestLike.test.concurrent;
        } else {
            return jestLike.test;
        }
    };

    const defineScenario = (
        scenarioTitle: string,
        scenarioFromStepDefinitions: ScenarioFromStepDefinitions,
        parsedScenario: ParsedScenario,
        only: boolean = false,
        skip: boolean = false,
        concurrent: boolean = false,
        timeout: number | undefined = undefined,
    ) => {
        const testFunction = getTestFunction(parsedScenario.skippedViaTagFilter, only, skip, concurrent);

        testFunction(scenarioTitle, () => {
            return scenarioFromStepDefinitions.steps.reduce((promiseChain, nextStep, index) => {
                const parsedStep = parsedScenario.steps[index];
                const stepArgument = parsedStep.stepArgument;
                const matches = matchSteps(
                    parsedStep.stepText,
                    scenarioFromStepDefinitions.steps[index].stepMatcher,
                );

                let matchArgs: string[] = [];

                if (matches && (matches as RegExpMatchArray).length) {
                    matchArgs = (matches as RegExpMatchArray).slice(1);
                }

                const args = [...matchArgs];

                if (stepArgument !== undefined && stepArgument !== null) {
                    args.push(stepArgument as string);
                }

                return promiseChain.then(() => {
                  return Promise.resolve()
                    .then(() => nextStep.stepFunction(...args))
                    .catch((error) => {
                        error.message = `Failing step: "${parsedStep.stepText}"\n\nStep arguments: ${JSON.stringify(args)}\n\nError: ${error.message}`;
                        throw error;
                    });
                });
            }, Promise.resolve());
        }, timeout);
    };

    const createDefineScenarioFunction = (
        featureFromStepDefinitions: FeatureFromStepDefinitions,
        parsedFeature: ParsedFeature,
        only: boolean = false,
        skip: boolean = false,
        concurrent: boolean = false,
    ) => {
        const defineScenarioFunction: DefineScenarioFunction = (
            scenarioTitle: string,
            stepsDefinitionFunctionCallback: StepsDefinitionCallbackFunction,
            timeout?: number,
        ) => {
            const scenarioFromStepDefinitions: ScenarioFromStepDefinitions = {
                title: scenarioTitle,
                steps: [],
            };

            featureFromStepDefinitions.scenarios.push(scenarioFromStepDefinitions);

            stepsDefinitionFunctionCallback({
                defineStep: createDefineStepFunction(scenarioFromStepDefinitions),
                given: createDefineStepFunction(scenarioFromStepDefinitions),
                when: createDefineStepFunction(scenarioFromStepDefinitions),
                then: createDefineStepFunction(scenarioFromStepDefinitions),
                and: createDefineStepFunction(scenarioFromStepDefinitions),
                but: createDefineStepFunction(scenarioFromStepDefinitions),
                pending: () => {
                    // Nothing to do
                },
            });

            const parsedScenario = parsedFeature.scenarios
                .filter((s) => s.title.toLowerCase() === scenarioTitle.toLowerCase())[0];

            const parsedScenarioOutline = parsedFeature.scenarioOutlines
                .filter((s) => s.title.toLowerCase() === scenarioTitle.toLowerCase())[0];

            const options = parsedFeature.options;

            scenarioTitle = processScenarioTitleTemplate(
                scenarioTitle,
                parsedFeature,
                options,
                parsedScenario,
                parsedScenarioOutline,
            );

            ensureFeatureFileAndStepDefinitionScenarioHaveSameSteps(
                options,
                parsedScenario || parsedScenarioOutline,
                scenarioFromStepDefinitions,
            );

            if (checkForPendingSteps(scenarioFromStepDefinitions)) {
                xtest(scenarioTitle, () => {
                    // Nothing to do
                }, undefined);
            } else if (parsedScenario) {
                defineScenario(
                    scenarioTitle,
                    scenarioFromStepDefinitions,
                    parsedScenario,
                    only,
                    skip,
                    concurrent,
                    timeout,
                );
            } else if (parsedScenarioOutline) {
                parsedScenarioOutline.scenarios.forEach((scenario) => {
                    defineScenario(
                        (scenario.title || scenarioTitle),
                        scenarioFromStepDefinitions,
                        scenario,
                        only,
                        skip,
                        concurrent,
                        timeout,
                    );
                });
            }
        };

        return defineScenarioFunction;
    };

    const createDefineScenarioFunctionWithAliases = (
        featureFromStepDefinitions: FeatureFromStepDefinitions,
        parsedFeature: ParsedFeature,
    ) => {
        const defineScenarioFunctionWithAliases = createDefineScenarioFunction(
            featureFromStepDefinitions,
            parsedFeature,
        );

        (defineScenarioFunctionWithAliases as DefineScenarioFunctionWithAliases).only = createDefineScenarioFunction(
            featureFromStepDefinitions,
            parsedFeature,
            true,
            false,
            false,
        );

        (defineScenarioFunctionWithAliases as DefineScenarioFunctionWithAliases).skip = createDefineScenarioFunction(
            featureFromStepDefinitions,
            parsedFeature,
            false,
            true,
            false,
        );

        (defineScenarioFunctionWithAliases as DefineScenarioFunctionWithAliases).concurrent =
            createDefineScenarioFunction(
                featureFromStepDefinitions,
                parsedFeature,
                false,
                false,
                true,
            );

        return defineScenarioFunctionWithAliases as DefineScenarioFunctionWithAliases;
    };

    const createDefineStepFunction = (scenarioFromStepDefinitions: ScenarioFromStepDefinitions) => {
        return (stepMatcher: string | RegExp, stepFunction: () => any) => {
            const stepDefinition: StepFromStepDefinitions = {
                stepMatcher,
                stepFunction,
            };

            scenarioFromStepDefinitions.steps.push(stepDefinition);
        };
    };

    return function defineFeature(
        featureFromFile: ParsedFeature,
        scenariosDefinitionCallback: ScenariosDefinitionCallbackFunction,
    ) {
        const featureFromDefinedSteps: FeatureFromStepDefinitions = {
            title: featureFromFile.title,
            scenarios: [],
        };

        const parsedFeatureWithTagFiltersApplied = applyTagFilters(featureFromFile);

        if (
            parsedFeatureWithTagFiltersApplied.scenarios.length === 0
                && parsedFeatureWithTagFiltersApplied.scenarioOutlines.length === 0
        ) {
            return;
        }

        jestLike.describe(featureFromFile.title, () => {
            scenariosDefinitionCallback(
                createDefineScenarioFunctionWithAliases(featureFromDefinedSteps, parsedFeatureWithTagFiltersApplied),
            );

            checkThatFeatureFileAndStepDefinitionsHaveSameScenarios(
                parsedFeatureWithTagFiltersApplied,
                featureFromDefinedSteps,
            );
        });
    };
};
