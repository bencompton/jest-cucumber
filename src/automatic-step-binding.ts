import { ParsedFeature, ParsedScenario, ParsedScenarioOutline } from './models';
import { matchSteps } from './validation/step-definition-validation';
import { StepsDefinitionCallbackFunction, defineFeature } from './feature-definition-creation';
import { generateStepCode } from './code-generation/step-generation';

const globalSteps: Array<{ stepMatcher: string | RegExp, stepFunction: () => any }> = [];

export const autoBindSteps = (features: ParsedFeature[], stepDefinitions: StepsDefinitionCallbackFunction[], registerStepsGlobally: boolean = true) => {
    const steps = registerStepsGlobally? globalSteps : [];
    const registerStep = (stepMatcher: string | RegExp, stepFunction: () => any) => steps.push({ stepMatcher, stepFunction });

    stepDefinitions.forEach((stepDefinitionCallback) => {
        stepDefinitionCallback({
            defineStep: registerStep,
            given: registerStep,
            when: registerStep,
            then: registerStep,
            and: registerStep,
            but: registerStep,
            pending: () => {
                // Nothing to do
            },
        });
    });

    const errors: string[] = [];

    features.forEach((feature) => {
        defineFeature(feature, (test) => {
            const scenarioOutlineScenarios = feature.scenarioOutlines
                .map((scenarioOutline) => scenarioOutline.scenarios[0]);

            const scenarios = [...feature.scenarios, ...scenarioOutlineScenarios];

            scenarios.forEach((scenario) => {
                test(scenario.title, (options) => {
                    scenario.steps.forEach((step, stepIndex) => {
                        const matches = steps
                            .filter((globalStep) => matchSteps(step.stepText, globalStep.stepMatcher));

                        if (matches.length === 1) {
                            const match = matches[0];

                            options.defineStep(match.stepMatcher, match.stepFunction);
                        } else if (matches.length === 0) {
                            const stepCode = generateStepCode(scenario.steps, stepIndex, false);
                            // tslint:disable-next-line:max-line-length
                            errors.push(`No matching step found for step "${step.stepText}" in scenario "${scenario.title}" in feature "${feature.title}". Please add the following step code: \n\n${stepCode}`);
                        } else {
                            const matchingCode = matches.map((match) => `${match.stepMatcher.toString()}\n\n${match.stepFunction.toString()}`);
                            errors.push(`${matches.length} step definition matches were found for step "${step.stepText}" in scenario "${scenario.title}" in feature "${feature.title}". Each step can only have one matching step definition. The following step definition matches were found:\n\n${matchingCode.join('\n\n')}`);
                        }
                    });
                });
            });
        });
    });

    if (errors.length) {
        throw new Error(errors.join('\n\n'));
    }
};
