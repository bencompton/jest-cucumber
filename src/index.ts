export { loadFeature, loadFeatures, parseFeature } from './parsed-feature-loading';
export { DefineStepFunction } from './feature-definition-creation';
export { setJestCucumberConfiguration } from './configuration';
export {
  generateCodeFromFeature,
  generateCodeWithSeparateFunctionsFromFeature,
} from './code-generation/generate-code-by-line-number';
export { StepsDefinitionCallbackFunction as StepDefinitions, IJestLike } from './feature-definition-creation';

import { createDefineFeature, IJestLike } from './feature-definition-creation';
import { createAutoBindSteps } from './automatic-step-binding';

const jestLike: IJestLike = {
  describe,
  test,
};

export const defineFeature = createDefineFeature(jestLike);
export const autoBindSteps = createAutoBindSteps(jestLike);
