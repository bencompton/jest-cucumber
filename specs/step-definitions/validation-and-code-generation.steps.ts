import { loadFeature, defineFeature, DefineStepFunction } from '../../src';
import { MockTestRunner } from '../utils/mock-test-runner/mock-test-runner';
import { MockStepDefinitions, wireUpMockFeature } from '../utils/wire-up-mock-scenario';
import {
    emptyStepDefinitions,
    featureWithMultipleScenarios,
    featureWithSingleScenario,
    featureWithSteplessScenario,
    steplessStepDefinitions,
    stepsForfeatureWithMultipleScenarios,
    stepsWithMismatchedSecondStep,
    stepsWithMissingStep,
    stepsWithScenariosOutOfOrder,
    stepsWithStepsOutOfOrder,
} from '../test-data/validation-and-code-generation';
import { defaultConfiguration, defaultErrorSettings, ErrorOptions, Options } from '../../src/configuration';

const feature = loadFeature('./specs/features/validation-and-code-generation.feature');

defineFeature(feature, (test) => {
    let mockTestRunner: MockTestRunner;
    let errorMessage = '';
    let options: Options;
    let stepDefinitions: MockStepDefinitions | null = null;
    let featureFile: string | null = null;

    beforeEach(() => {
        mockTestRunner = new MockTestRunner();
        errorMessage = '';
        options = { ...defaultConfiguration };
        options.errors = { ...defaultErrorSettings };
    });

    const givenScenariosMustMatchFeatureFileIsEnabled = (given: DefineStepFunction) => {
        given(/`scenariosMustMatchFeatureFile` is (enabled|disabled)/, (enabledOrDisabled) => {
            (options.errors as ErrorOptions).scenariosMustMatchFeatureFile =
                enabledOrDisabled === 'enabled';
        });
    };

    const givenScenariosMustMatchFeatureFileIsDisabled = givenScenariosMustMatchFeatureFileIsEnabled;

    const givenStepsMustMatchFeatureFileIsEnabled = (given: DefineStepFunction) => {
        given(/`stepsMustMatchFeatureFile` is (enabled|disabled)/, (enabledOrDisabled) => {
            (options.errors as ErrorOptions).stepsMustMatchFeatureFile =
                enabledOrDisabled === 'enabled';
        });
    };

    const givenAllowScenariosNotInFeatureFileIsEnabled = (given: DefineStepFunction) => {
        given(/`allowScenariosNotInFeatureFile` is (enabled|disabled)/, (enabledOrDisabled) => {
            (options.errors as ErrorOptions).allowScenariosNotInFeatureFile =
                enabledOrDisabled === 'enabled';
        });
    };

    const givenAllowScenariosNotInFeatureFileIsDisabled = givenAllowScenariosNotInFeatureFileIsEnabled;

    const givenStepsMustMatchFeatureFileIsDisabled = givenStepsMustMatchFeatureFileIsEnabled;

    const whenIRunMyJestCucumberTests = (when: DefineStepFunction) => {
        when('I run my Jest Cucumber tests', () => {
            try {
                wireUpMockFeature(mockTestRunner,
                    featureFile as string,
                    stepDefinitions,
                    options,
                );
            } catch (err) {
                errorMessage = err.message;
            }
        });
    };

    const thenIShouldSeeAValidationErrorAndGeneratedCode = (then: DefineStepFunction) => {
        then(/I should (see|not see) a validation error \/ generated code/, (seeOrNotSee) => {
            if (seeOrNotSee === 'see') {
                expect(errorMessage).toBeTruthy();
                expect(errorMessage).toMatchSnapshot();
            } else {
                expect(errorMessage).toBeFalsy();
            }
        });
    };

    const thenIShouldNotSeeAValidationErrorAndGeneratedCode = thenIShouldSeeAValidationErrorAndGeneratedCode;

    const thenIShouldSeeAValidationError = (then: DefineStepFunction) => {
        then(/I should (see|not see) a validation error/, (seeOrNotSee) => {
            if (seeOrNotSee === 'see') {
                expect(errorMessage).toBeTruthy();
                expect(errorMessage).toMatchSnapshot();
            } else {
                expect(errorMessage).toBeFalsy();
            }
        });
    };

    const thenIShouldNotSeeAValidationError = thenIShouldSeeAValidationError;

    const andAScenarioInAFeatureFileHasNoStepDefinitions = (and: DefineStepFunction) => {
        and('a scenario in a feature file has no step definitions', () => {
            featureFile = featureWithSingleScenario;
            stepDefinitions = emptyStepDefinitions;
        });
    };

    const andTheStepOrderDiffersBetweenTheFeatureAndStepDefinitions = (and: DefineStepFunction) => {
        and('the step order differs between the feature / step definitions', () => {
            featureFile = featureWithSingleScenario;
            stepDefinitions = stepsWithStepsOutOfOrder;
        });
    };

    const andTheStepCountDiffersBetweenTheFeatureAndStepDefinitions = (and: DefineStepFunction) => {
        and('the step count differs between the feature / step definitions', () => {
            featureFile = featureWithSingleScenario;
            stepDefinitions = stepsWithMissingStep;
        });
    };

    const andIHaveAScenarioWhereTheStepMatcherForTheSecondStepDoesntMatchTheStep = (and: DefineStepFunction) => {
        and('I have a scenario where the step matcher for the second step doesn\'t match the step', () => {
            featureFile = featureWithSingleScenario;
            stepDefinitions = stepsWithMismatchedSecondStep;
        });
    };

    const andThereIsAnExtraScenarioInMyStepDefinitionsNotInMyFeatureFile = (and: DefineStepFunction) => {
        and('there is an extra scenario in my step definitions not in my feature file', () => {
            featureFile = featureWithSingleScenario;
            stepDefinitions = stepsForfeatureWithMultipleScenarios;
        });
    };

    test('Enabled and scenario missing step definitions', ({ given, and, when, then }) => {
        givenScenariosMustMatchFeatureFileIsEnabled(given);
        andAScenarioInAFeatureFileHasNoStepDefinitions(and);
        whenIRunMyJestCucumberTests(when);
        thenIShouldSeeAValidationErrorAndGeneratedCode(then);
    });

    test('Disabled and scenario missing step definitions', ({ given, and, when, then }) => {
        givenScenariosMustMatchFeatureFileIsDisabled(given);
        andAScenarioInAFeatureFileHasNoStepDefinitions(and);
        whenIRunMyJestCucumberTests(when);
        thenIShouldNotSeeAValidationErrorAndGeneratedCode(then);
    });

    test('Enabled and scenario filtered via tag filter', ({ given, and, when, then }) => {
        givenScenariosMustMatchFeatureFileIsEnabled(given);
        andAScenarioInAFeatureFileHasNoStepDefinitions(and);

        and('that scenario is filtered via a tag filter', () => {
            options.tagFilter = 'not @my-tag';
        });

        whenIRunMyJestCucumberTests(when);
        thenIShouldNotSeeAValidationErrorAndGeneratedCode(then);
    });

    test('Scenario order is the same', ({ given, and, when, then }) => {
        givenScenariosMustMatchFeatureFileIsEnabled(given);

        and('a set of scenarios is ordered the same in the feature / step definitions', () => {
            featureFile = featureWithMultipleScenarios;
            stepDefinitions = stepsForfeatureWithMultipleScenarios;
        });

        whenIRunMyJestCucumberTests(when);
        thenIShouldNotSeeAValidationErrorAndGeneratedCode(then);
    });

    test('Scenario order is different', ({ given, and, when, then }) => {
        givenScenariosMustMatchFeatureFileIsEnabled(given);

        and('a set of scenarios is ordered the differently in the feature / step definitions', () => {
            featureFile = featureWithMultipleScenarios;
            stepDefinitions = stepsWithScenariosOutOfOrder;
        });

        whenIRunMyJestCucumberTests(when);
        thenIShouldNotSeeAValidationErrorAndGeneratedCode(then);
    });

    test('Enabled and step order / count are the same', ({ given, and, when, then }) => {
        givenStepsMustMatchFeatureFileIsEnabled(given);

        and('the step definitions exactly match the feature file', () => {
            featureFile = featureWithMultipleScenarios;
            stepDefinitions = stepsForfeatureWithMultipleScenarios;
        });

        whenIRunMyJestCucumberTests(when);
        thenIShouldNotSeeAValidationErrorAndGeneratedCode(then);
    });

    test('Enabled and step order is different', ({ given, and, when, then }) => {
        givenStepsMustMatchFeatureFileIsEnabled(given);
        andTheStepOrderDiffersBetweenTheFeatureAndStepDefinitions(and);
        whenIRunMyJestCucumberTests(when);
        thenIShouldSeeAValidationErrorAndGeneratedCode(then);
    });

    test('Enabled and step count is different', ({ given, and, when, then }) => {
        givenStepsMustMatchFeatureFileIsEnabled(given);
        andTheStepCountDiffersBetweenTheFeatureAndStepDefinitions(and);
        whenIRunMyJestCucumberTests(when);
        thenIShouldSeeAValidationErrorAndGeneratedCode(then);
    });

    test('Disabled and step order is different', ({ given, and, when, then }) => {
        givenStepsMustMatchFeatureFileIsDisabled(given);
        andTheStepOrderDiffersBetweenTheFeatureAndStepDefinitions(and);
        whenIRunMyJestCucumberTests(when);
        thenIShouldNotSeeAValidationErrorAndGeneratedCode(then);
    });

    test('Disabled and step count is different', ({ given, and, when, then }) => {
        givenStepsMustMatchFeatureFileIsDisabled(given);
        andTheStepCountDiffersBetweenTheFeatureAndStepDefinitions(and);
        whenIRunMyJestCucumberTests(when);
        thenIShouldNotSeeAValidationErrorAndGeneratedCode(then);
    });

    // tslint:disable-next-line: max-line-length
    test('Enabled and a step in the step definitions doesn\'t match the step in the feature', ({ given, and, when, then }) => {
        givenStepsMustMatchFeatureFileIsEnabled(given);
        andIHaveAScenarioWhereTheStepMatcherForTheSecondStepDoesntMatchTheStep(and);
        whenIRunMyJestCucumberTests(when);
        thenIShouldSeeAValidationErrorAndGeneratedCode(then);
    });

    // tslint:disable-next-line: max-line-length
    test('Disabled and a step in the step definitions doesn\'t match the step in the feature', ({ given, and, when, then }) => {
        givenStepsMustMatchFeatureFileIsDisabled(given);
        andIHaveAScenarioWhereTheStepMatcherForTheSecondStepDoesntMatchTheStep(and);
        whenIRunMyJestCucumberTests(when);
        thenIShouldNotSeeAValidationErrorAndGeneratedCode(then);
    });

    test('Scenario with no steps', ({ given, and, when, then }) => {
        givenStepsMustMatchFeatureFileIsDisabled(given);

        and('I have a scenario with no steps', () => {
            featureFile = featureWithSteplessScenario;
            stepDefinitions = steplessStepDefinitions;
        });

        whenIRunMyJestCucumberTests(when);
        thenIShouldNotSeeAValidationErrorAndGeneratedCode(then);
    });

    test('Enabled and step definitions have an extra scenario', ({ given, and, when, then }) => {
        givenAllowScenariosNotInFeatureFileIsEnabled(given);
        andThereIsAnExtraScenarioInMyStepDefinitionsNotInMyFeatureFile(and);
        whenIRunMyJestCucumberTests(when);
        thenIShouldNotSeeAValidationError(then);
    });

    test('Disabled and step definitions have an extra scenario', ({ given, and, when, then }) => {
        givenAllowScenariosNotInFeatureFileIsDisabled(given);
        andThereIsAnExtraScenarioInMyStepDefinitionsNotInMyFeatureFile(and);
        whenIRunMyJestCucumberTests(when);
        thenIShouldSeeAValidationError(then);
    });
});
