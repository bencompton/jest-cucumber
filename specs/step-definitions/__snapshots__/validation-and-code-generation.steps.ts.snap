// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation and code generation Disabled and step definitions have an extra scenario 1`] = `"No scenarios found in feature file that match scenario title \\"Doing some more stuff.\\""`;

exports[`Validation and code generation Enabled and a step in the step definitions doesn't match the step in the feature 1`] = `
"Expected step #2 in scenario \\"Doing some stuff\\" to match \\"I do some stuff\\". Try adding the following code:

when('I do some stuff', () => {

});"
`;

exports[`Validation and code generation Enabled and scenario missing step definitions 1`] = `
"Feature file has a scenario titled \\"Doing some stuff\\", but no match found in step definitions. Try adding the following code:

test('Doing some stuff', ({ given, when, then }) => {
	given('I did some stuff', () => {

	});

	when('I do some stuff', () => {

	});

	then('I should have done some stuff', () => {

	});
});"
`;

exports[`Validation and code generation Enabled and step count is different 1`] = `
"Scenario \\"Doing some stuff\\" has 3 step(s) in the feature file, but 2 step definition(s) defined. Try adding the following code:

test('Doing some stuff', ({ given, when, then }) => {
	given('I did some stuff', () => {

	});

	when('I do some stuff', () => {

	});

	then('I should have done some stuff', () => {

	});
});"
`;

exports[`Validation and code generation Enabled and step order is different 1`] = `
"Expected step #2 in scenario \\"Doing some stuff\\" to match \\"I do some stuff\\". Try adding the following code:

when('I do some stuff', () => {

});

Expected step #3 in scenario \\"Doing some stuff\\" to match \\"I should have done some stuff\\". Try adding the following code:

then('I should have done some stuff', () => {

});"
`;
